src/runners/baseRunner/BaseRunner.ts

```diff
@@ -42,52 +38,63 @@ export abstract class BaseRunner implements ITestRunner {
 
-                socket.on('data', (buffer: Buffer) => {
-                    data = data.concat(iconv.decode(buffer, launchConfiguration.encoding || 'utf8'));
-                    const index: number = data.lastIndexOf(os.EOL);
-                    if (index >= 0) {
-                        this.testResultAnalyzer.analyzeData(data.substring(0, index + os.EOL.length));
-                        data = data.substring(index + os.EOL.length);
-                    }
-                });
+            socket.on('data', (buffer: Buffer) => {
+                data = data.concat(iconv.decode(buffer, launchConfiguration.encoding || 'utf8'));
+                const index: number = data.lastIndexOf(os.EOL);
+                if (index >= 0) {
+                    this.testResultAnalyzer.analyzeData(data.substring(0, index + os.EOL.length));
+                    data = data.substring(index + os.EOL.length);
+                }
+            });
 
-                socket.on('close', (had_error: boolean) => {
-                    if (had_error) {
-                        return reject(new Error(`Launcher failed to run.`));
-                    }
-                    if (this.isCanceled) {
-                        return resolve([]);
-                    }
-                    if (data.length > 0) {
-                        this.testResultAnalyzer.analyzeData(data);
-                    }
-                    return resolve(this.testResultAnalyzer.feedBack());
-                });
+            socket.on('error', (err: Error) => {
+                throw err;
             });
 
             this.server.on('error', (err: Error) => {
-                return reject(err);
+                throw err;
             });
-
-            this.launchTests(launchConfiguration);
         });
+
+        const uri: Uri = Uri.parse(this.tests[0].location.uri);
+        logger.verbose(`Launching with the following launch configuration: '${JSON.stringify(launchConfiguration, null, 2)}'\n`);
+
+        return await debug.startDebugging(workspace.getWorkspaceFolder(uri), launchConfiguration).then(async (success: boolean) => {
+            if (!success) {
+                this.tearDown();
+                return [];
+            }
+
+            return await new Promise<ITestResult[]>((resolve: (result: ITestResult[]) => void): void => {
+                this.disposables.push(
+                    debug.onDidTerminateDebugSession((session: DebugSession): void => {
+                        if (launchConfiguration.name === session.name) {
+                            this.tearDown();
+                            if (data.length > 0) {
+                                this.testResultAnalyzer.analyzeData(data);
+                            }
+                            return resolve(this.testResultAnalyzer.feedBack());
+                        }
+                    }),
+                );
+            });
+        }, ((reason: any): any => {
+            logger.error(`${reason}`);
```

Comment:
reason

Reply:
Seems I need to refine the logger's APIs... Will do it in the following PRs